### PR TITLE
WEBCORE-20: Remove cut dependency on Changelog

### PR DESCRIPTION
**Pull request by mxkyb**
## Description

Will create two pull requests:

### Against main
Will raise the version in `${project}/package.json`, package-lock.json and `${project}/CHANGELOG.md`

### Against release
Will create a release candidate to merge against release.

## Related
https://github.com/zattoo/web/pull/4788

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -4,9 +4,18 @@ inputs:
...
     description: 'User based token with write permission'
     required: true
-  labels:
+  rc-labels:
     description: 'List of labels to add to release candidate'
     required: false
+  main-labels:
+    description: 'List of labels to add to version raise'
+    required: false
+  new-version:
```
Comment (gotbahn) [inline action.yml:13]: `-` → `_`

File: src/github-api.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ * We need to manually create the commit because of file size limits of the octokit api,
+ * wich we hit with package.lock.json
+ *
+ * More explanation:
+ * https://git-scm.com/book/en/v2/Git-Internals-Git-Objects
+ * https://octokit.github.io/rest.js/v18
+ */
+import * as github from '@actions/github';
+
+const BLOB_MODE_FILE = '100644';
```
Comment (gotbahn) [inline src/github-api.js:11]: what's that?

File: action.yml [hunk 3]
```diff
@@ -4,9 +4,18 @@ inputs:
...
+    description: 'Version of the next release'
+    required: true
+  project:
+    description: 'Project to cut'
+    required: true
```
Comment (nitzanashi) [inline action.yml:18]: why so many labels?

File: src/github-api.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+let octokit = null;
```
Comment (nitzanashi) [inline src/github-api.js:13]:
```suggestion
let octokit;

File: src/index.js [hunk 5]
```diff
@@ -23,156 +23,267 @@ const exit = (message, exitCode) => {
...
+    const rcLabels = core.getMultilineInput('rc-labels', {required: false});
+    const versionRaiseLabels = core.getMultilineInput('main-labels', {required: false});
+    const project = core.getInput('project', {required: true});
+    const newVersion = core.getInput('new-version', {required: true});
+
+    init(token);
 
-    if (!foundSomething) {
-        exit('No release candidates were found', 0);
+    if (newVersion.split('.').length !== 3) {
```
Comment (gotbahn) [inline src/index.js:229]: no this is not enough, it also should support both semver and zver

File: src/github-api.js (hunk 2 above)
Comment (mxkyb) [reply src/github-api.js:11]: its the code git uses for mark blobs as files

**Comment by nitzanashi** [comment]
@mxkyb any explanation on what is going on would be helpful it super hard to understand it with +31k changes 
How do we cut now? XD

**Comment by mxkyb** [comment]
haha sorry :) i thought, "review is best served cold" :)

**Comment by mxkyb** [comment]
(i am going to write something)

File: src/index.js (hunk 5 above)
Comment (mxkyb) [reply src/index.js:229]: hm, why? aren't they both `[major version].[year][week].[patch]` and `[major].[minor].[patch]`

**Comment by mxkyb** [comment]
@gotbahn @nitzanashi added some description into the readme. when review is finished i will also update the wiki page

File: src/index.js [hunk 6]
```diff
@@ -23,156 +23,267 @@ const exit = (message, exitCode) => {
...
-
-        const newVersions = getNewVersions(project, changelogBefore, changelogAfter);
-
-        if (!isEmpty(newVersions)) {
-            await Promise.all(newVersions.map((version) => cut(project, version)));
-        }
+    // sequencial: commit hashes need to be in order
+    await updatePackageLock();
+    await updatePackageJson();
+    await updateChangelog();
```
[the remark this answers is not in the document]
Comment (mxkyb) [reply src/index.js:130]: not with `Promise.all`, but we could use a single commit instead, yes. will look into it later

File: src/index.js [hunk 7]
```diff
@@ -23,156 +23,267 @@ const exit = (message, exitCode) => {
...
+            owner,
+            repo,
+            branch: rcBranch,
+            sha: baseSha,
+        }),
+    ]);
+
+    let changelogEntries = '';
+
+    const updateChangelog = async () => {
```
[the remark this answers is not in the document]
Comment (mxkyb) [reply src/index.js:176]: can be outside, will do

**Review by gotbahn — COMMENTED** [review]
what's left here?

**Comment by mxkyb** [comment]
@gotbahn batching of commits, will do that today

**Comment by mxkyb** [comment]
@nitzanashi @gotbahn 

Specification updated here: https://zattoo2.atlassian.net/wiki/spaces/WEB/pages/3443261460/Project+Release#Cutting
and here: https://github.com/zattoo/cactus/blob/de3ff0ce13fd2b0f4a950b7b08d03c37828d4678/README.md

**Review by gotbahn — DISMISSED** [review]

File: src/github-api.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    branch,
+    sha,
+}) => {
+    await octokit.rest.git.createRef({
+        owner,
+        repo,
+        ref: `refs/heads/${branch}`,
+        sha,
+    });
+};
```
Comment (nitzanashi) [inline src/github-api.js:37]:
```suggestion
export const createBranch = async (data) => {
    const {
        branch,
        ...rest
    } = data;
    
    await octokit.rest.git.createRef({
        ...rest,
        ref: `refs/heads/${branch}`,
    });
};
```

File: src/index.js [hunk 9]
```diff
@@ -23,156 +15,274 @@ const exit = (message, exitCode) => {
     process.exit(exitCode);
 };
 
-const getNewVersions = (project, changelogBefore, changelogAfter) => {
-    let newVersions = [];
+export const validateVersion = (previousVersion, nextVersion) => {
+    if (previousVersion === nextVersion) {
+        exit('Version musst be different', 1);
```
Comment (nitzanashi) [inline src/index.js:20]:
```suggestion
        exit('Version must be different', 1);
```

File: src/index.js [hunk 10]
```diff
@@ -23,156 +15,274 @@ const exit = (message, exitCode) => {
     process.exit(exitCode);
 };
 
-const getNewVersions = (project, changelogBefore, changelogAfter) => {
-    let newVersions = [];
+export const validateVersion = (previousVersion, nextVersion) => {
```
Comment (nitzanashi) [inline src/index.js:18]: export?

File: src/index.js [hunk 11]
```diff
@@ -23,156 +15,274 @@ const exit = (message, exitCode) => {
...
-    const repo = repository.name;
-    const owner = repository.full_name.split('/')[0];
+const editPackageLock = ({
+    rawPackageLock,
+    nextVersion,
+    project,
+}) => {
+    const packageLock = JSON.parse(rawPackageLock);
+
+    packageLock.packages[`projects/${project}`].version = nextVersion;
```
Comment (nitzanashi) [inline src/index.js:97]: I don't like that the structure of our repo is hardcoded into this Action, would it be possible to pass this information via the job itself?

File: src/index.js [hunk 12]
```diff
@@ -23,156 +15,274 @@ const exit = (message, exitCode) => {
...
+    });
 
-    if (isEmpty(changelogs)) {
-        exit('No changelog changes', 0);
-    }
+    await github.createPullRequest({
+        owner,
+        repo,
+        title: `Next ${project}`,
+        body: `Bump version`,
```
Comment (nitzanashi) [inline src/index.js:149]:
```suggestion
        body: 'Bump version',
```

File: src/index.js [hunk 13]
```diff
@@ -23,156 +15,274 @@ const exit = (message, exitCode) => {
...
+        rawChangelog: files.changelog,
+    });
+
+    await github.createCommit({
+        owner,
+        repo,
+        branch: rcBranch,
+        paths: {
+            changelog: paths.changelog,
+            serviceFile: `projects/${project}/.release-servcie`,
```
Comment (nitzanashi) [inline src/index.js:200]:
```suggestion
            serviceFile: `projects/${project}/.release-service`,
```
what is this?

File: src/index.js (hunk 11 above)
Comment (nitzanashi) [reply src/index.js:97]: We can set default value and have option to pass it in the job

File: src/index.js (hunk 13 above)
Comment (nitzanashi) [reply src/index.js]: I read the docs but couldn't understand it, can you explain it for dummis?

Comment (mxkyb) [reply src/index.js]: We can only open a pull request in when there are file changes. the release branch and the rc branch are both forked from the main branch, so we need some change that allows us to merge rc into release. normally we use changelog for that (like here https://github.com/zattoo/web/pull/4780/files), but we want to be independent from the changelog. so this service file just contains a random unique hash that allows us to remove changelog in the future

File: src/index.js (hunk 11 above)
Comment (mxkyb) [reply src/index.js]: the default value is taken from the yaml file, checked it in web-shadow and it works :)

**Review by gotbahn — APPROVED** [review]

**Pull request merged**